CIPs/0016.md

```diff
@@ -0,0 +1,89 @@
+- enable simple construction of QRs by code from arbitrary celo transactions
+
+## Proposed Solution
+
+### Syntax
+Celo protocol URLs contain "celo" in their schema (protocol) part and are constructed as follows:
+
+    request                 = "celo:" [ prefix "-" ] target [ "/" endpoint ] [ "?" parameters ]
+    prefix                  = "pay" / "request" / "call" / "send"
+    
+    target                  = E164_PHONE_NUMBER / registry_id / celo_address
+    registry_id             = "StableToken" / "GoldToken" / "LockedGold" / ...
+    celo_address            = ( "0x" 40*40HEXDIG )
+    
+    endpoint                = token / function
+    token                   = "cUSD" / "CELO" / ...
+    function                = function_name "(" [ function_args ] ")"
+    function_name           = STRING
+    function_args           = function_arg *( "," function_arg )
+    function_arg            = ( NUMBER / STRING ) [ ":" TYPE ]
+
+    parameters              = parameter *( "&" parameter )
+    parameter               = key "=" value
+    key                     = "value" / "gas" / "gasPrice" / "feeCurrency" / "gatewayFee" / "gatewayFeeRecipient" / "localAmount" / "localCurrencyCode" / ...
+    value                   = NUMBER / STRING / celo_address
+
+### Semantics
+
+`TYPE` is a standard ABI type name, as defined in [Ethereum Contract ABI specification](https://solidity.readthedocs.io/en/develop/abi-spec.html). `STRING` is a URL-encoded unicode string of arbitrary length, where delimiters and the percentage symbol (`%`) are mandatorily hex-encoded with a `%` prefix.
+
+`E164_PHONE_NUMBER` is a standard E164 universal phone number, as defined by the International Telecommunication Union in [E164 Number Specification](https://www.itu.int/rec/T-REC-E.164).
+
+`prefix` is used for differentiating between user and contract `target`s and actions. `pay` and `request` are for user targets and intended for use with phone numbers or addresses and the `token` endpoint. `call` & `send` are for contract targets and intended for use with the registry/addresses and the `function` endpoint. 
+
+`token` is provided as an endpoint to more concisely encode celo protocol-aware ERC20 transfers that would otherwise be represented via a `function` endpoint.
+
+`registry_id` is provided to simplify and shorten celo protocol-aware contract interactions (to be eventually translated by the application to a registry-specified 40 byte address). 
+
+`function_args` have an optional `TYPE` because, in some cases, the application may have knowledge of the ABIs deeming the solidity types unnecessary. These can most likely be omitted when used with a celo registry contract via a `registry_id` target.
+
+`localAmount` is intended for use with `localCurrencyCode` and specifies a value in a locale-specific denomination (to be eventually translated by the application to a celo-specific denomination). Resolving `value` with `localAmount` is discretionary.
```

Comment:
Curious about this decision, can you elaborate on why `localAmount`/`localCurrencyCode` were chosen rather than translating to a `value` client side and encoding that in the QR code instead?

Reply:
The constructor may not have access to the relevant exchange rate APIs and/or a merchant may want to denominate an item's price in a local denomination without being subject to exchange rate fluctuations.